pkg/natdiscovery/natdiscovery.go

```diff
@@ -0,0 +1,176 @@
+
+package natdiscovery
+
+import (
+	"crypto/rand"
+	"math/big"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+
+	v1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/types"
+)
+
+type Interface interface {
+	Run(stopCh <-chan struct{}, natResultCh chan RemoteEndpointNAT) error
+	AddEndpoint(endpoint *types.SubmarinerEndpoint)
+	RemoveEndpoint(endpointName string)
+	SetReadyChannel(readyChannel chan *RemoteEndpointNAT)
+}
+
+type udpWriteFunction func(b []byte, addr *net.UDPAddr) (int, error)
+type findSrcIPFunction func(destinationIP string) (string, error)
+
+type natDiscovery struct {
+	sync.Mutex
+	localEndpoint   *types.SubmarinerEndpoint
+	remoteEndpoints map[string]*RemoteEndpointNAT
+	requestCounter  uint64
+	serverUDPWrite  udpWriteFunction
+	findSrcIP       findSrcIPFunction
+	serverPort      int
+	readyChannel    chan *RemoteEndpointNAT
+}
+
+func (nd *natDiscovery) SetReadyChannel(readyChannel chan *RemoteEndpointNAT) {
+	nd.readyChannel = readyChannel
+}
+
+func New(localEndpoint *types.SubmarinerEndpoint) (Interface, error) {
+	return newNatDiscovery(localEndpoint)
+}
+
+func newNatDiscovery(localEndpoint *types.SubmarinerEndpoint) (*natDiscovery, error) {
+	port, err := extractNATDiscoveryPort(localEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	requestCounter, err := randomRequestCounter()
+	if err != nil {
+		return nil, err
+	}
+
+	return &natDiscovery{
+		localEndpoint:   localEndpoint,
+		serverPort:      port,
+		remoteEndpoints: map[string]*RemoteEndpointNAT{},
+		findSrcIP:       findPreferredSourceIP,
+		requestCounter:  requestCounter,
+	}, nil
+}
+
+func randomRequestCounter() (uint64, error) {
+	n, err := rand.Int(rand.Reader, big.NewInt(1000))
+	if err != nil {
+		return 0, errors.Wrapf(err, "generating random request counter")
+	}
+
+	return n.Uint64(), nil
+}
+
+var errorNoNatDiscoveryPort = errors.New("no nat discovery port in endpoint")
+
+func extractNATDiscoveryPort(endpoint *types.SubmarinerEndpoint) (int, error) {
+	if endpoint.Spec.Ports == nil {
+		return 0, errorNoNatDiscoveryPort
+	}
+
+	if portString, ok := endpoint.Spec.Ports[v1.NATDiscoveryPort]; ok {
+		port, err := strconv.Atoi(portString)
+		if err != nil {
+			return 0, errors.Wrapf(err, "error trying to extract port number from %q in %s",
+				portString, v1.NATDiscoveryPort)
+		}
+
+		return port, nil
+	}
+
+	return 0, errorNoNatDiscoveryPort
+}
+
+func (nd *natDiscovery) Run(stopCh <-chan struct{}, natResultCh chan RemoteEndpointNAT) error {
+	err := nd.runListener(stopCh)
+	if err != nil {
+		return err
+	}
+
+	go wait.Until(func() {
+		nd.checkEndpointList()
+	}, time.Second, stopCh)
+
+	return nil
+}
+
+func (nd *natDiscovery) AddEndpoint(endpoint *types.SubmarinerEndpoint) {
+	remoteNAT := newRemoteEndpointNAT(endpoint)
+
+	// support a remote cluster endpoint which still hasn't implemented this protocol
```

Comment:
If AddEndpoint is called periodically, we will be validating NAT-discovery repeatedly right?

Reply:
Oh we need to check that case, good point.

Reply:
done. Now we will only replace it if the endpoint has changed (public ip changed or nat settings changed, for example)